### PR TITLE
Added Dark Mode Feature to thecodingtrain Website 

### LINKED ISSUE #1504: Add Dark Mode Feature to thecodingtrain Website
# Dark Mode Feature Proposal

**Description:**
As a user of thecodingtrain website, I often find myself browsing late at night or in low-light environments where a dark mode would be more comfortable for my eyes. I propose adding a dark mode feature to the website to enhance user experience and accessibility.

**Benefits:**
- Improves user experience, especially in low-light environments.
- Provides greater accessibility options for users with sensitivity to bright light.
- Aligns with current web design trends and user preferences.

**Additional Notes:**
- This feature could be made toggleable via a button in the website's header or through a settings menu.
- browser preferences to enable easy switching between light and dark modes.

### Dark Mode
<details>
<summary>Desktop</summary>

Dark Mode - Desktop

</details>

<details>
<summary>Small Screen (Mobile/Tablet)</summary>

Dark Mode - Small Screen

</details>

### Light Mode
<details>
<summary>Desktop</summary>

Light Mode - Desktop

</details>

<details>
<summary>Small Screen (Mobile/Tablet)</summary>

Light Mode - Small Screen

</details>

Feel free to provide any feedback or suggestions for improving this feature.

Once I complete the user preferences part, I will proceed to create a pull request. @shiffman

### PR DESCRIPTION
**Pull request by Yashasewi**
# Dark Mode Implementation

This pull request introduces the implementation of the dark mode feature for the thecodingtrain.com website, as discussed in #1504.

## Changes

- Implemented dark mode using CSS variables and theming classes
- Created a dedicated `theme.js` component to change CSS variables for light and dark modes
- Updated existing styles to reference the appropriate CSS variables for theming
- Addressed instances of hard-coded colors and updated them to use CSS variables

### CSS Variables and Theming

The `theme.js` file defines two sets of CSS variables: one for the light mode and another for the dark mode. These variables control various colors and styles throughout the website, including text, backgrounds, and UI elements.

```Javascript
/* Light Mode Variables */
const lightTheme = [
  '--text-color: #000000',
  '--gray-dark: #3c3c3c',
  '--gray-mid: #666666',
  ......
];

/* Dark Mode */
const darkTheme = [
  '--text-color: #f0f0f0',
  '--background-color: #1e1d20',
  '--gray-dark: #1f1f1f',
  '--gray-mid: #b3b3b3',
  ......
];
```

The `theme.js` component manages the application of the appropriate CSS variables based on the user's preference (light or dark mode). It also persists the user's preference in localStorage for consistency across sessions.

### Existing Styles Updates

Existing CSS styles have been updated to reference the appropriate CSS variables, ensuring that colors and visual elements adapt seamlessly to the selected theme.

## Potential Issues

-  While updating the styles to use CSS variables, there might be some instances where colors were hard-coded or missed. These instances could potentially lead to inconsistencies or unintended visual effects in the dark mode. Further testing and review from the team would be beneficial to identify and address such cases.
-  Additionally, the icons in the footer cannot be changed by simply adjusting the color, so we might need to use new icons or find an alternative solution to ensure they are visible and consistent with the dark mode.
- Regarding the gradient present on the home page's homescene, I couldn't find a suitable alternative that works well with the dark mode. I have left this part for the design team to decide on a new gradient or alternative design element that looks good in the dark mode.

## Next Steps

- Address any feedback or suggestions from the review process
- Conduct accessibility audits to ensure compliance with WCAG guidelines
- Identify and address any remaining instances of hard-coded colors or inconsistencies
- Please review the changes and provide your feedback. I'm happy to address any concerns or make further adjustments as needed.

Resolves #1504

@shiffman @runemadsen @jasontheillustrator

### CODE REVIEW

**Comment by shiffman** [comment]
Thank you again @Yashasewi I am so excited to have this, it is amazing work! (And thank you for your patience, I know it can be frustrating to have contributions linger for a long time!).

I have reviewed the deploy preview and am happy with this! Does anyone have any remaining technical comments before I merge?

Also, I believe I found a bug, one some pages in dark model there are still "light" sections, for example:

https://deploy-preview-1554--codingtrain.netlify.app/tracks/data-and-apis-in-javascript

Screen Shot 2024-04-20 at 9 42 23 AM

**Comment by dipamsen** [comment]
other inconsistencies/bugs:

Footer social links - twitch icon has a different background
image

Track card - language/topic tags not visible
image

Track video header - section with light background, language/topic tags not visible
image

VideoSection - timestamps not visible, section with light background, nebula link not visible
image

VideoInfo + PassengerShowcasePanel - links/description/credits not visible, passenger showcase header not visible, showcase name not visible
image

(same for "Related Challenges" section)

PassengerShowcaseForm - dark text
image

**Comment by Yashasewi** [comment]
Hi @shiffman and @dipamsen,

Thank you for taking the time to review the dark mode implementation and for your valuable feedback. I appreciate you pointing out the issues and inconsistencies you've encountered. (No worries, I know these things take time! 😄)

Regarding the bug you mentioned, @shiffman, where some pages in dark mode still have "light" sections (e.g., `deploy-preview-1554--codingtrain.netlify.app/tracks/data-and-apis-in-javascript`), I will investigate the issue and ensure that the dark mode styles are applied consistently across all pages.

As for the other inconsistencies/bugs reported:

1. **Footer social links - icon issue?**
  - The current SVG icons used in the footer do not seem to support the dark mode styles effectively. Since these icons are in SVG format, simply adjusting the color might not work as expected. I think we need to consider using different icon sets or formats that support dark mode natively, or explore alternative solutions to ensure the icons are visible and consistent in both light and dark modes. 

> what should we do about the icons ? @shiffman any suggestions...

2. **Other Issues**
  - Thank you for catching the other issues. I think I missed some pages on the website. But sit back and relax, I will fix the other issues soon.

@dipamsen, if you could share the specific URLs where you encountered the issues, it would be extremely helpful for me to locate and fix them easily.


I will work on addressing them and update the pull request accordingly. If you encounter any other inconsistencies or issues, please feel free to share them as well.
 I appreciate your collaboration, support, and understanding in ensuring a seamless integration of the dark mode feature.

**Comment by dipamsen** [comment]
@Yashasewi i have updated my comment to include the required urls to pages on the deploy preview site

**Comment by Yashasewi** [comment]
Hi @shiffman and @dipamsen,

- [x] Track card - language/topic tags not visible ([url](https://deploy-preview-1554--codingtrain.netlify.app/tracks))
- [x] Track video header - section with light background, language/topic tags not visible ([url](https://deploy-preview-1554--codingtrain.netlify.app/tracks/code-programming-with-p5-js))
- [x] VideoSection - timestamps not visible, section with light background ([url](https://deploy-preview-1554--codingtrain.netlify.app/tracks/code-programming-with-p5-js))
- [x] PassengerShowcaseForm - dark text ([url](https://deploy-preview-1554--codingtrain.netlify.app/guides/passenger-showcase-guide))
- [x] VideoInfo + PassengerShowcasePanel - links/description/credits not visible, passenger showcase header not visible, showcase name not visible ([url](https://deploy-preview-1554--codingtrain.netlify.app/tracks/code-programming-with-p5-js/code/3-conditionals/2-bouncing))
- [ ] Footer Social Links - Icon Issue
- As @dipamsen mentioned that the current SVG icons used in the footer do not handle dark mode styles effectively.
- Proposed Solution: I can change the icons to new icons that support dark mode switching, ensuring consistent visibility and styling.


Regarding the footer social links, if I have the necessary permissions, I will proceed with changing the icons to new ones that support dark mode switching. This should resolve the icon visibility and styling issues in both light and dark modes.

I have fixed most of the bugs found during your review. However, if you encounter any additional issues or inconsistencies, please do not hesitate to mention them.

**Comment by dipamsen** [comment]
amazing! one last place that i missed mentioning, is the challenge page which has some white areas, as well as dark text ([example](https://deploy-preview-1554--codingtrain.netlify.app/challenges/182-apollonian-gasket))

Also, the search results for tracks ([url](https://deploy-preview-1554--codingtrain.netlify.app/tracks/lang/all/topic/drawing)) and the date of video on the challenge page ([url](https://deploy-preview-1554--codingtrain.netlify.app/challenges/)) could be light text?

**Comment by Yashasewi** [comment]
Hii @dipamsen 
Fixed most of the issues.  Please check it and if you find any bugs or inconsistency please mention here.

**Comment by dipamsen** [comment]
Looks great!!!  After a check I found few more places:
- About Page ([url](https://deploy-preview-1554--codingtrain.netlify.app/about))
- Track video page ([url](https://deploy-preview-1554--codingtrain.netlify.app/tracks/code-programming-with-p5-js/code/1-intro/5-errors)) - track stops/time stamps label not visible on hover 
  image
- Challenge page ([url](https://deploy-preview-1554--codingtrain.netlify.app/challenges/134-heart-curve)) - parts/time stamps label not visible (on active and hover)
image

**Comment by Yashasewi** [comment]
Hii @dipamsen Please check it and if you find any bugs or inconsistency .... mention here.

**Comment by fturmel** [comment]
Hey everyone! I'm a bit late to the dark mode discussion, apologies. Great job @Yashasewi, looks like we're getting close!

## Theme toggle

I have some recommendations on the implementation.

### General approach
It's counter-intuitive, but with a React framework that pre-renders static content at build time and needs to hydrate, using Context and React state is not the ideal way to go. It will cause a kind of FOUC where the user will see the wrong theme flash on first render, which we can observe in your current solution.

Here's a better approach, which is how the React team implement their own dark mode on [react.dev](https://react.dev/):
* Inject some JS code in the head or at the top of the body (it _should_ block rendering) that will handle the localStorage + setting a "dark" class in the html tag if needed. This should probably live in the Layout or Head component of this project. Tailwind has a nice and simple example in their docs: https://tailwindcss.com/docs/dark-mode#supporting-system-preference-and-manual-selection

* Always render _both_ toggle buttons in JXS, but hide the one that shouldn't be currently displayed with CSS selectors that leverage the presence of the "dark" class in the html element.

* The theme toggle buttons can just add/remove the "dark" class and update the localStorage value directly. Again, I would reference the Tailwind link previously mentioned.

This should fix the FOUC and will not cause React hydration warnings. Let me know if you need help. I could take this on in the PR directly or do a follow up later.

### Theme CSS variables in CSS file

It would be more in line with the project to move the CSS variables to an actual global CSS file (variables.css) instead of having them in JS. You can then dedupe/merge the light theme variables and add a `html.dark` selector for the dark theme.

### Button semantic

Maybe I'm missing something, but could we not use a button element instead of making a span behave like one? It would be nice to add a pointer cursor, hover state and title too.

## Misc

### Social icons in footer

This new icon for Nebula doesn't look like their logo, the rotation is wrong and the star has rounded points. Maybe try `io5/IoStarSharp` with `transform: scaleY(-1);` instead?

We can also get rid of the old SVG icons if they're not used anywhere else.

### Train illustration

Am I the only one that finds the wheels on the train illustration particularly odd-looking against a dark background?

CleanShot 2024-04-28 at 17 00 24@2x


### Styles

Have we properly tested on mobile, and also looked for regressions on the light theme yet?

I have not had the time to do so, but one thing that jumped at me was the gray background in the footer being gone:

| Before | After |
| -| - |
| CleanShot 2024-04-28 at 16 52 00@2x | CleanShot 2024-04-28 at 16 51 46@2x |

**Comment by fturmel** [comment]
@Yashasewi sounds good, happy to help. It might be easier for me to also move the CSS variables as I make the toggle implementation changes. I should have some time to get this done at some point tomorrow.

There's a potential for conflicts if we work on this at the same time because things are intertwined a bit, so make sure you push any pending local commits on your end.

**Comment by fturmel** [comment]
@Yashasewi All good, I think you can pick things up from here!

**Comment by Yashasewi** [comment]
@fturmel Thank you so much for your contribution to the dark mode implementation. It has significantly improved the overall user experience and brought us closer to delivering a polished dark mode feature.

There is one more issue that I've been trying to fix but haven't been successful yet. I would really appreciate your help with this if possible.

The issue is related to the search/filter input field. When you try to search or filter on specific pages like [thecodingtrain url](https://deploy-preview-1554--codingtrain.netlify.app/tracks) or the challenges page, the input text color becomes dark, which doesn't contrast well with the dark background and tends to blend in.

I've tried different approaches to fix this UI bug using CSS, but I haven't been able to resolve it successfully. If you could take a look and provide a solution or fix it , it would be extremely helpful.

Here's a video demonstrating the issue: [video link](https://www.loom.com/share/61e4e7ed135840c79f22136d24186c46?sid=9cba94ac-7c4c-44b2-840a-d91eef703ef7)
Also  in image 
image

**Comment by Yashasewi** [comment]
@fturmel Thank you again for your assistance with the search/filter input field issue in dark mode. I really appreciate your efforts in finding a solution.

You're correct that the react-select component's documentation on manual CSS styling is lacking, which made it difficult for me to identify the correct class names to target. Despite my attempts to style it from the CSS side, I couldn't achieve the desired result.

However, I noticed that when we type anything in the field, the text input still appears in a dark color, which doesn't provide enough contrast against the dark background.

 Could you please take a look at this remaining issue and see if we can further improve the visibility and contrast of the text input within the search/filter field?

image

**Comment by shiffman** [comment]
@all-contributors add @Yashasewi for code

**Comment by shiffman** [comment]
This is so fantastic and such a heroic effort, thank you to all involved! I'm ready to merge this! As far as I can tell everything seems resolved, but let me know if I should hold off if there are continued refinements needed? Also tagging in @runemadsen in case you want to have one last look in terms of design? huge ❤️ for this!

**Comment by fturmel** [comment]
@shiffman Not _quite_ there yet, but very close. We need a bit more QA, fix the Nebula icon in footer and fix the background color regression in the footer light theme.

Let me put this in draft mode temporarily.

**Comment by fturmel** [comment]
I just went through the whole site and made some additional improvements. Ready for a last look and merge imo.

**Comment by Yashasewi** [comment]
@fturmel I had made similar local changes to address some of the remaining issues, but after you went through the entire site and made these commits, I don't need to push those local commits anymore. The improvements and fixes you've implemented cover the areas I had planned to work on.

I was also testing this site on mobile devices, and the bugs I found were related to the code examples and the navigation not closing on mobile when activated. However, you have fixed both of these issues with your recent commits. I couldn't find any other issues or problems on mobile devices.

I believe we can safely merge this pull request into the main branch. If someone found any issue or bug or inconsistency, please mention here.

**Comment by fturmel** [comment]
@Yashasewi Sorry I stole your thunder! It made sense to just power through and fix the small issues as I was QA'ing locally.

I feel bad because I did spend time writing detailed comments before so you could have some direction and work it out yourself 🤦

I'll be more patient next time! And again, good job on this PR.

**Comment by Yashasewi** [comment]
@fturmel 😂 There's nothing to feel bad about at all! Working alongside an experienced contributor like yourself has been an incredible learning opportunity for me.
How about we add dark mode to the Nature of Code site next? That would be a great follow-up project .

**Comment by shiffman** [comment]
Thank you, wow! I see that the checks are not passing, let me know if there's something I should do or check into! @Yashasewi I love the idea of dark mode for nature of code, but the site and its design are still quite a bit in flux so it probably would make sense to wait until it's more settled and final. Feel free to poke around the repo and look at the issues and other discussion! I welcome contributions!

**Comment by fturmel** [comment]
Alright, all checks are green again. There must've been an outage / glitch on Netlify's end a few days ago.

**Comment by dipamsen** [comment]
I am noticing that foreground text color on certain sections of the light theme site has been changed from `#3c3c3c` (`--gray-dark`) to `#000000`, is this intended?

| Before | After |
|--------|-------|
| image | image  |

**Comment by dipamsen** [comment]
Also, the color pair of foreground and background on the [Guides](https://deploy-preview-1554--codingtrain.netlify.app/guides) page (as well as Showcase, About) may be difficult to read due to low color contrast ([does not conform to WCAG AA](https://webaim.org/resources/contrastchecker/?fcolor=7C3B82&bcolor=1E1D20)), maybe we can use a different (lighter) shade/color theme for these pages for the light theme?

**Comment by dipamsen** [comment]
While I am listing the changes (regressions), the social icons seems to have been changed, presumably because the original SVGs which were used could not be dynamically used for both light and dark theme. Currently icons from `react-icons` library are used instead.

| Before | After | After |
|--------|--------|-------|
| image | image | image |


P.S. Is it okay to just rotate a star and call it the nebula logo?

https://github.com/CodingTrain/thecodingtrain.com/blob/77cc4945abb603ecb7048fb666daa134dfeaf2f1/src/components/Footer.js#L102-L107

File: src/components/PassengerShowcaseForm.js
Comment (dipamsen) [inline src/components/PassengerShowcaseForm.js:370]: This change makes the label "Social media permission" a part of the option _"The Cafe Car: ..."_, which makes it so that clicking on the label "Social media permission" selects that option, which may not be intended.

**Comment by Yashasewi** [comment]
@dipamsen, thanks for raising those points.

Regarding the color updates, I've changed a few colors by adding opacity and reverted to the old purple color.

As for the Nebula icon, you're right that it's just a rotated star design. The `react-icons` library doesn't have an official Nebula icon, and this rotated star seems to be their official branding representation:

nebula-navbar-light@2x-95459d40

I agree it's not an ideal solution, but it was the best approach we could find for now to ensure the icon works well in both light and dark modes.

File: src/components/PassengerShowcaseForm.js
Comment (fturmel) [reply src/components/PassengerShowcaseForm.js:370]: Good catch, that was unintended. I made that quick edit to get rid of the linter warning (labels without an associated input are technically invalid markup). We can revert this change and I'll properly address the problem when I tackle #1550.

**Comment by Yashasewi** [comment]
I was thinking about adding syntax highlighting to the "Maintaining the Website" guide page. This would improve the readability .



Additionally, I found a bug on mobile devices when viewing the [Maintaining the Website Guide page](https://deploy-preview-1554--codingtrain.netlify.app/guides/maintain-website-guide).



This issue occurs only on Chromium-based browsers such as Google Chrome, DuckDuckGo, and Brave. Unfortunately, I don't have access to an iPhone to confirm if the issue is also present in Safari. Also this bug is not on the mozilla firefox. 

Please let me know your thoughts on implementing syntax highlighting for the code . I'm happy to work on these improvements and welcome any additional feedback or suggestions you might have.

**Comment by fturmel** [comment]
@Yashasewi Maybe you can open new PRs / issues for these so we can concentrate on dark mode related concerns here?

**Comment by shiffman** [comment]
I have no preference about the social icons and am glad to go with what is easiest to implement and maintain!

I agree with @fturmel, I like your additional ideas @Yashasewi but let's work on them in separate issues / branches and focus on dark mode here!

**Comment by fturmel** [comment]
> I have no preference about the social icons and am glad to go with what is easiest to implement and maintain!

The current state of the PR is the easiest and simplest (react-icons without circles). The work is done, it's dark mode compatible without additional efforts, and the brand logos will be up-to-date.

Are we ready to merge? @jasontheillustrator @runemadsen did you want to chime in on the feature?

**Comment by Yashasewi** [comment]
@fturmel @shiffman 
Thanks I will open other issues for those. I guess than  we are ready to merge .

**Comment by shiffman** [comment]
Fantastic, let's give this ~24 hours for any last comments and then I will go ahead and merge!

**Comment by dipamsen** [comment]
@fturmel @Yashasewi https://github.com/CodingTrain/thecodingtrain.com/pull/1554#issuecomment-2094274899 ?

**Comment by Yashasewi** [comment]
> @fturmel @Yashasewi https://github.com/CodingTrain/thecodingtrain.com/pull/1554#issuecomment-2094274899 ?

Hii @dipamsen now if you check it again they look quite similar to me . But sometimes it's harder to differentiate colours for me . So please let me know if you think we need to change the colour a bit more .

| Before | After |
|:------:|:-----:|
| Before | After |



And In most cases the text colour has been replaced by `--text-color: #000000cb;` variable.  Also changed the opacity of text colour. https://github.com/CodingTrain/thecodingtrain.com/pull/1554/commits/74964aa03cf790da9dcc5d19f2d2dbe8d426d95c

**Review by dipamsen — APPROVED** [review]
Looks good. Apparently originally some texts were complete black whereas some other places it was a lighter opacity. (unsure whether intentional). Currently it is consistent for all copy.

Should be ready for merge! (unless there are any more comments)

**Comment by shiffman** [comment]
Heeeeeeeere weeeeeeeeeee goooooooooooooo! Amazing work @Yashasewi!

**Comment by Yashasewi** [comment]
Let's gooooooooooooooooooooooooooooooo
lets-go-lets-go-marvel
We did it . 🤜🤛

**Pull request merged**
